File: supabase/functions/_shared/fast2sms.ts
// Fast2SMS API client for Supabase Edge Functions
// https://docs.fast2sms.com/reference/authorization

export function isConfigured(): boolean {
  const apiKey = Deno.env.get("FAST2SMS_API_KEY");
  return !!apiKey && apiKey.trim().length > 0;
}

/**
 * Sends an SMS message via Fast2SMS API using authorization header.
 * @param phone 10-digit mobile number or format with country code
 * @param message Text message content
 * @param otp Optional 6-digit OTP code if using route "otp"
 */
export async function sendSmsMessage(
  phone: string,
  message: string,
  otp?: string,
): Promise<void> {
  const apiKey = Deno.env.get("FAST2SMS_API_KEY") ?? "";
  const route = (Deno.env.get("FAST2SMS_ROUTE") ?? "otp").toLowerCase();

  if (!apiKey) {
    console.log("Fast2SMS API key not set, skipping SMS send.");
    return;
  }

  let phoneClean = phone.replace(/\D/g, "");
  if (phoneClean.length > 10 && phoneClean.startsWith("91")) {
    phoneClean = phoneClean.slice(2);
  }

  if (phoneClean.length !== 10) {
    throw new Error(`Invalid phone number for Fast2SMS: ${phone}`);
  }

  const url = "https://www.fast2sms.com/dev/bulkV2";

  let payload: Record<string, unknown>;

  if (route === "otp" && otp) {
    payload = {
      route: "otp",
      variables_values: otp,
      numbers: phoneClean,
    };
  } else {
    payload = {
      route: "q",
      message,
      language: "english",
      flash: 0,
      numbers: phoneClean,
    };
  }

  const res = await fetch(url, {
    method: "POST",
    headers: {
      "authorization": apiKey,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(payload),
  });

  if (!res.ok) {
    const errorBody = await res.text();
    console.error(`Fast2SMS API error: ${res.status}`, errorBody);
    throw new Error(`Failed to send SMS via Fast2SMS: ${res.status} ${errorBody}`);
  }
}

File: backend/app/infrastructure/fast2sms.py
"""Fast2SMS SMS Client — https://docs.fast2sms.com/reference/authorization"""

from app import config
from app.infrastructure.http_client import resilient_post
from app.infrastructure.logging import get_logger

log = get_logger("turbofix.fast2sms")
FAST2SMS_BULK_URL = "https://www.fast2sms.com/dev/bulkV2"


def is_configured() -> bool:
    return bool(getattr(config, "FAST2SMS_API_KEY", "").strip())


async def send_sms_message(phone: str, message: str, otp_code: str = "") -> dict:
    phone_clean = "".join(c for c in phone if c.isdigit())[-10:]
    api_key = getattr(config, "FAST2SMS_API_KEY", "").strip()
    if not api_key:
        log.info("fast2sms.not_configured", phone=phone_clean, otp=otp_code)
        return {"status": "simulated", "detail": "FAST2SMS_API_KEY not configured"}

    route = getattr(config, "FAST2SMS_ROUTE", "otp").strip().lower()
    payload = (
        {"route": "otp", "variables_values": otp_code, "numbers": phone_clean}
        if route == "otp" and otp_code
        else {"route": "q", "message": message, "language": "english", "flash": 0, "numbers": phone_clean}
    )

    try:
        resp = await resilient_post(
            FAST2SMS_BULK_URL,
            json=payload,
            headers={"authorization": api_key, "Content-Type": "application/json"},
        )
        res_data = resp.json()
        log.info("fast2sms.sent", phone=phone_clean, route=route, result=res_data)
        return {"status": "sent", "response": res_data}
    except Exception as exc:
        log.error("fast2sms.error", phone=phone_clean, error=str(exc))
        raise

